File: 10344 - 23 out of 5/10344.cpp
// Nombre: Esteban David Muñoz Bautista
// Código: 898233

#include <bits/stdc++.h>
using namespace std;

bool encontrado = false;

bool solveRec(vector<int> &numeros, int valor, int cont) {
    bool ans = false;

    if (cont == 0) {
        if (valor == 23) {
            encontrado = true;
            ans = true;
        }
    } 
    else {
        int sumita = 0;
        for (int x : numeros) sumita += x;

        if (!((valor * sumita < 23 && valor + sumita < 23) ||
              (valor - sumita > 23 && valor * sumita > 23))) {

            int i = 0;
            while (i < 5 && !encontrado) {
                if (numeros[i] != 0 && !encontrado) {
                    int num = numeros[i];
                    numeros[i] = 0;

                    int j = 0;
                    while (j < 3 && !encontrado) {
                        if (j == 0 && !encontrado)
                            ans = solveRec(numeros, valor + num, cont - 1);
                        else if (j == 1 && !encontrado)
                            ans = solveRec(numeros, valor - num, cont - 1);
                        else if (j == 2 && !encontrado)
                            ans = solveRec(numeros, valor * num, cont - 1);

                        j = j + 1;
                    }

                    numeros[i] = num;
                }
                i = i + 1;
            }
        }
    }

    return ans;
}

bool solveInicial(vector<int> numeros) {
    encontrado = false;
    bool ans = false;

    int i = 0;
    while (i < 5 && !encontrado) {
        int num = numeros[i];
        numeros[i] = 0;
        solveRec(numeros, num, 4);
        numeros[i] = num;
        i = i + 1;
    }

    if (encontrado) ans = true;
    return ans;
}

int main() {
    vector<int> numeros(5);
    bool seguir = true;

    while (seguir) {
        int suma = 0;
        for (int i = 0; i < 5; i++) {
            cin >> numeros[i];
            suma = suma + numeros[i];
        }

        if (suma == 0) {
            seguir = false;
        } else {
            bool posible = solveInicial(numeros);
            if (posible) cout << "Possible\n";
            else cout << "Impossible\n";
        }
    }

    return 0;
}

File: 10344 - 23 out of 5/10344.py
# Nombre: Esteban David Muñoz Bautista
# Código: 898233

from sys import stdin

encontrado  = False

def solve(numeros, valor, cont):
    global encontrado 
    ans = False
    if(cont == 0):
        if(valor == 23):
            encontrado  = True
            ans = True
    else:
        sumita = sum(numeros)
        if not ((valor * sumita <23 and valor + sumita < 23) or (valor - sumita > 23 and valor * sumita >23)): 
            i = 0
            while(i < 5 and not encontrado ):
                if(numeros[i] != 0 and not encontrado ):
                    num = numeros[i]
                    numeros[i] = 0
                    j = 0
                    while(j < 3 and not encontrado ):
                        if(j == 0 and not encontrado ):
                            ans = solve(numeros, valor + num, cont - 1)
                        elif(j == 1 and not encontrado ):
                            ans = solve(numeros, valor - num, cont - 1)
                        elif(j == 2 and not encontrado ):
                            ans = solve(numeros, valor * num, cont - 1)
                        j += 1
                    numeros[i] = num
                i += 1
    return ans


def main():
    global encontrado 
    numeros = list(map(int, stdin.readline().split()))
    while(sum(numeros) > 0):
        encontrado  = False        
        i = 0
        while(i < 5 and not encontrado ):
            num = numeros[i]
            numeros[i] = 0
            solve(numeros, num, 4)
            numeros[i] = num
            i += 1
        
        if(encontrado ):
            print("Possible")
        else:
            print("Impossible")
        
        numeros = list(map(int, stdin.readline().split()))

main()